### PR TITLE
Fix correctly setting Find-Scopes when opening Find/Replace-Dialog #1819

### PR DESCRIPTION
**Pull request by Wittmaxi**
This PR correctly disables SearchOptions.GLOBAL in the FindReplaceLogic object of the Dialog if a multiline-selection was selected while opening the Find/Replace-Dialog.

This PR also provides a unit-test for this behavior.

See https://github.com/eclipse-platform/eclipse.platform.ui/issues/1819

### CODE REVIEW

**Comment by Wittmaxi** [comment]
This PR is branched from https://github.com/eclipse-platform/eclipse.platform.ui/pull/1805 since I needed the changes to the Testing Infrastructure I made in that PR in order to test this feature. The diff will look far cleaner, once #1805 is merged

File: tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java [hunk 1]
```diff
@@ -426,6 +429,19 @@ public void testActivateWholeWordsAndSearchForTwoWords() {
...
 
+	@Test
+	public void testActivateDialogWithSelectionActive() {
+		openTextViewer("text" + System.lineSeparator() + "text" + System.lineSeparator() + "text");
+		fTextViewer.setSelectedRange(5, 9);
+		openFindReplaceDialogForTextViewer(false);
+
+		assertFalse(dialog.globalRadioButton.getSelection());
+		dialog.findCombo.setText("text");
+		dialog.replaceAllButton.notifyListeners(SWT.Selection, null);
```
Comment (HeikoKlare) [inline tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java]: Please use `select(dialog.replaceAllButton())`

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java [hunk 2]
```diff
@@ -68,7 +68,7 @@ public interface IFindReplaceLogic {
 
 	/**
 	 * {@return whether incremental search may be performed by the
-	 * find/replace-logic based on the currently active options}
+	 * find/replace-logic based on the currently active options.}
```
Comment (HeikoKlare) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java]: Please revert this unrelated change

**Review by HeikoKlare — COMMENTED** [review]
Looks good, thank you!
Just two minor comments.

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java (hunk 2 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java]: Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/d2b76269413f3578e8a605d2410049c3801569be

File: tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java (hunk 1 above)
Comment (Wittmaxi) [reply tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java]: Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/d2b76269413f3578e8a605d2410049c3801569be

it is worth noting that select does more than just call the listeners - it actively sets a selection on the replaceAllButton.
Since the button doesn't care about it's "selection", I believe that this is fine.

File: tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java [hunk 3]
```diff
@@ -426,6 +429,20 @@ public void testActivateWholeWordsAndSearchForTwoWords() {
...
+	@Test
+	public void testActivateDialogWithSelectionActive() {
+		openTextViewer("text" + System.lineSeparator() + "text" + System.lineSeparator() + "text");
+		fTextViewer.setSelectedRange(5, 9);
+		openFindReplaceDialogForTextViewer(false);
+
+		assertFalse(dialog.globalRadioButton.getSelection());
+		dialog.findCombo.setText("text");
+		select(dialog.replaceAllButton);
+		runEventQueue();
```
Comment (HeikoKlare) [inline tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java]: Why should it be necessary to run the event queue here? The test fails on Linux despite this call. I think what is missing in general is to run the event queue once the action to open the find/replace dialog has been invoked (like is introduced in #1192).

File: tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java [hunk 4]
```diff
@@ -426,6 +429,20 @@ public void testActivateWholeWordsAndSearchForTwoWords() {
...
+
+		assertEquals(fTextViewer.getDocument().get(), "text" + System.lineSeparator() + System.lineSeparator());
```
Comment (HeikoKlare) [inline tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java]: Please switch expected and actual.

File: tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java [hunk 5]
```diff
@@ -426,6 +429,20 @@ public void testActivateWholeWordsAndSearchForTwoWords() {
 		assertFalse(dialog.wholeWordCheckBox.getEnabled());
 	}
 
+	@Test
+	public void testActivateDialogWithSelectionActive() {
+		openTextViewer("text" + System.lineSeparator() + "text" + System.lineSeparator() + "text");
+		fTextViewer.setSelectedRange(5, 9);
```
Comment (HeikoKlare) [inline tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java]: The range should properly consider the length of `lineSeparator()`.

File: tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java (hunk 3 above)
Comment (HeikoKlare) [reply tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java]: Root cause of the test failure will be fixed by #1836.

File: tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java (hunk 4 above)
Comment (Wittmaxi) [reply tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java]: I have chosen to use Hamcrest-Matchers instead in order to avoid this confusion. I could imagine this being considered inconsistent as this is the only place I use hamcrest in this test-class. Since the new FindReplaceDialogTest (introduced in the Find/Replace-Overlay-PR) uses Hamcrest, I don't believe this is a problem.

File: tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java (hunk 3 above)
Comment (Wittmaxi) [reply tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java]: Will #1836 really fix the test? From my understanding, the busy-indicator won't make this thread wait for performReplaceAll.

What we need for this test is a good way to know when the performReplaceAll-method is done running - no matter which thread it is in.

What is the correct procedure for right now?

Comment (HeikoKlare) [reply tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java]: It will fix the test, because the test fails because of a `NullPointerException` when executing "replace all", as the active shell is still `null` (missing GTK event processing). Doing a `runEventQueue()` after initializing the dialog and before doing the "replace all" would also do the trick, but is rather a hack that will not be necessary after merging #1836.

**Review by HeikoKlare — APPROVED** [review]
Great, thank you! Failing test is unrelated and documented in #926.
